euphonic/spectra.py

```diff
@@ -821,16 +819,44 @@ def __getitem__(
             self, item: Union[Integral, slice, Sequence[Integral], np.ndarray]
     ):  # noqa: F811
         self._validate_item(item)
-        init_kwargs = {
-            self._spectrum_data_name(): self._get_spectrum_data()[item, :],
-            "x_tick_labels": self.x_tick_labels,
-            "metadata": self._get_item_metadata(item)
-                       } | self._get_bin_kwargs()
 
         if isinstance(item, Integral):
-            return self._item_type(**init_kwargs)
+            spectrum = self._item_type.__new__(self._item_type)
+        else:
+            spectrum = self.__new__(type(self))
+
+        self._set_item_data(spectrum, item)
 
-        return type(self)(**init_kwargs)
+        spectrum.x_tick_labels = self.x_tick_labels
+        spectrum.metadata = self._get_item_metadata(item)
+
+        return spectrum
+
+    def _set_item_data(
+            self,
+            spectrum: Spectrum,
+            item: Union[Integral, slice, Sequence[Integral], np.ndarray]
+    ) -> None:
+        """Write axis and spectrum data from self to Spectrum
+
+        This is intended to set attributes on a 'bare' Spectrum created with
+        __new__() from the parent SpectrumCollection.
+        """
+
+        for axis in self._bin_axes:
+            setattr(spectrum, f"_{axis}_data",
+                    getattr(self, f"_{axis}_data").copy())
+            setattr(spectrum, f"_internal_{axis}_data_unit",
+                    getattr(self, f"_internal_{axis}_data_unit"))
+            setattr(spectrum, f"{axis}_data_unit",
+                    getattr(self, f"{axis}_data_unit"))
+
+        setattr(spectrum, self._spectrum_raw_data_name(),
+                self._get_raw_spectrum_data()[item, :].copy())
+        setattr(spectrum, f"_internal_{self._spectrum_data_name()}_unit",
+                self._get_internal_spectrum_data_unit())
+        setattr(spectrum, f"{self._spectrum_data_name()}_unit",
+                self._get_spectrum_data_unit())
```

Comment:
Is the first one right? The fact it uses `spectrum_raw_data_name` rather than `spectrum_data_name` makes me concerned this will get visually confusing.

Reply:
Copying the raw data is the _whole point_ of doing this. When we access the actual "spectrum data" (i.e. `y_data` in 1D or `z_data` in 2D) it does some expensive unit-handling. Here we just copy strings and numpy arrays.

Reply:
I don't object to that, I just think that when reading the code quickly, it'll be easy to miss that that's a different structure.

Reply:
Perhaps `self._raw_spectrum_data_name` would parse better than `_spectrum_raw_data_name`. And it's more consistent with `_get_raw_spectrum_data()`.

Reply:
That would certainly help. 

I don't know whether a blank line might also just visually separate that as being different to the ones below. I know they're doing something as a block, but I just get the subtle feeling that if I were modifying this code, I'd skip past that difference.